### ISSUE TITLE
[BUG] LB service included a space in the random string for its name

### ISSUE THREAD

**Issue by yaocw2020** [harvester/harvester#2122]
**Rancher Server Setup**
 
* Rancher version: v2.6.4-rc9
* Installation option (Docker install/Helm Chart): docker
   
 * If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc):
 * Proxy/Cert Details:
 
 **Information about the Cluster**
 
* Kubernetes version: 1.22.7+rke2r1
* Cluster Type (Local/Downstream): downstream
   
* If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): harvester node driver
 
**User Information**
 
* What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom): tested with admin
 
**Describe the bug** when creating an LB through a workload, the service name had a ` ` in it, preventing the service from coming up. (not always reproducible)
 
**To Reproduce**
 
* deploy a cluster with rke2 harvester node driver (cloud provider enabled)
 * deploy a workload on the cluster with an IPAM LB defined, using the `pool` option
 * wait for service to come up
 
 **Result** service is unable to deploy, character in name not allowed:
 
 <img alt="Screen Shot 2022-03-18 at 10 34 36 AM" width="599" src="https://user-images.githubusercontent.com/16691014/159063644-b588df46-6288-481c-a7d8-843d9d12e0b8.png">
 **Expected Result** the character(s) appended to a string by rancher should always be from the allowed set.

**Additional context**
 
* I have brought up over 50 LBs with this node driver/test but only hit this specific issue 1x so it is likely hard to reproduce.
 * not sure if this is exclusive to harvester, or could happen with any provider (I suspect the latter)

**Comment by yaocw2020** [comment]
Copy from https://github.com/rancher/rancher/issues/36956

**Comment by yaocw2020** [comment]
The suffix is a 32 bits checksum which could be represented as a seven- or eight-digit hexadecimal number. If it's a seven-digit hexadecimal number, the statement ```fmt.Sprintf("%8x", digest)``` will add a space at the beginning of the returned string. The correct statement is ```fmt.Sprintf("%08x", digest)``` which will add the 0 character rather than space.
``` golang
digest := crc32.ChecksumIEEE([]byte(name + string(service.UID)))
suffix := fmt.Sprintf("%8x", digest)
name += suffix
```

**Comment by yaocw2020** [comment]
Test plan:
- Deploy a guest cluster enabling the harvester cloud provider
- Create an Nginx deployment
- Create numbers of related load balancer services until one is always pending. Checking the cloud provider logs, we could see the logs like
```
E0404 19:10:08.299203   70133 controller.go:310] error processing service default/nginx-lb-1 (will retry): failed to ensure load balancer: create loadbalancer default/kubernetes-default-nginx-lb-1- 72beb2a failed, error: LoadBalancer.loadbalancer.harvesterhci.io "kubernetes-default-nginx-lb-1- 72beb2a" is invalid: metadata.name: Invalid value: "kubernetes-default-nginx-lb-1- 72beb2a": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```
- Replace the harvester cloud provider deployment image as the personal image `armarny/harvester-cloud-provider:issue2122`. The pending service should be normal now.

**Comment by TachunLin** [comment]
Verified **fixed** on harvester `v1.0.1` with Rancher `v2.6.4` after replacing cloud provider image to `armarny/harvester-cloud-provider:issue2122`

Close this issue.

### Result
After replacing cloud provider image to `armarny/harvester-cloud-provider:issue2122`. 

1. It can make the `pending` load balacner (lb-28) which have space in random string back to `running` status
  ![image](https://user-images.githubusercontent.com/29251855/163141931-d646799b-faf1-4610-b9ad-cd72c4932b88.png)

  ```
  E0413 09:07:15.977374 1 controller.go:307] error processing service default/lb-28 (will retry): failed to ensure load balancer: LoadBalancer.loadbalancer.harvesterhci.io "rke2-v1234-default-lb-28- c426c07" is invalid: metadata.name: Invalid value: "rke2-v1234-default-lb-28- c426c07": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
  I0413 09:07:15.977647 1 event.go:291] "Event occurred" object="default/lb-28" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: LoadBalancer.loadbalancer.harvesterhci.io \"rke2-v1234-default-lb-28- c426c07\" is invalid: metadata.name: Invalid value: \"rke2-v1234-default-lb-28- c426c07\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')"
  ```
  
  ![image](https://user-images.githubusercontent.com/29251855/163153295-6cb39526-33c9-4c70-9844-cd51227a6634.png)


2. Can route to the back to normal load balancer (lb-28) correctly  
  ![image](https://user-images.githubusercontent.com/29251855/163153519-5ddc9307-4ef9-47eb-b38f-1bb6e6f9054f.png)

### Test Information
* Test Environment: 1 node harvester on local kvm machine
* Harvester version: v1.0.1
* Rancher version: v2.6.4
* RKE2 version: v1.23.4
* OS image: ubuntu focal 

### Verify Steps
1. Deploy harvester and set the pool CIDR values in settings
  ![image](https://user-images.githubusercontent.com/29251855/160443207-2637d05b-f9e0-4a19-be9e-3b5dc69ddc94.png)
2. Deploy rancher + connect harvester via virtualization management
3. Deploy a harvester node driver cluster with cloud provider enabled
4. Deploy a nginx server deployment
5. Open kube shell in the RKE2 cluster 
6. Execute the following shell snippet to create bunch of load balancer in once 
  ```
  for i in {7..30}
  do
     kubectl expose deployment nginx --port=80 --type=LoadBalancer --name="lb-$i"
     sleep 2
  done
  ```
7. Check whether exists the `pending` load balancer in Service Discovery -> Services
  ![image](https://user-images.githubusercontent.com/29251855/163141931-d646799b-faf1-4610-b9ad-cd72c4932b88.png)
8. If yes, check the `cloud provider pods` logs in Workload -> Pods
  ![image](https://user-images.githubusercontent.com/29251855/163141868-59ac1d6f-8499-4a1b-92c6-22bdf61509cd.png)
  ```
  E0413 09:07:15.977374 1 controller.go:307] error processing service default/lb-28 (will retry): failed to ensure load balancer: LoadBalancer.loadbalancer.harvesterhci.io "rke2-v1234-default-lb-28- c426c07" is invalid: metadata.name: Invalid value: "rke2-v1234-default-lb-28- c426c07": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
  I0413 09:07:15.977647 1 event.go:291] "Event occurred" object="default/lb-28" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: LoadBalancer.loadbalancer.harvesterhci.io \"rke2-v1234-default-lb-28- c426c07\" is invalid: metadata.name: Invalid value: \"rke2-v1234-default-lb-28- c426c07\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')"
  ```
#### Next we need to replace the cloud provider image version
10. Open cloud provider application
  ![image](https://user-images.githubusercontent.com/29251855/163143183-65151e0c-5898-43a8-b5ab-1a5cea49e42b.png)

11. Edit yaml of the cloud-provider deployment
  ![image](https://user-images.githubusercontent.com/29251855/163143915-ab0cb85a-0966-4cbb-8b52-f7f75af64983.png)
  
12. Replace the image with `armarny/harvester-cloud-provider:issue2122` and save
  ![image](https://user-images.githubusercontent.com/29251855/163144157-135bf172-6238-49ec-a7c2-594092628b37.png)

13. Remove the `Running` pod
  ![image](https://user-images.githubusercontent.com/29251855/163145416-20648da8-6402-48fc-bf4d-4bed29a48ce5.png)
  ![image](https://user-images.githubusercontent.com/29251855/163145712-87fcc845-8705-4f59-b5b5-2af1bf011910.png)

14. Replace cloud provider to customized image
  ![image](https://user-images.githubusercontent.com/29251855/163145847-8357736d-3a9a-4cba-b5d5-38c7ab76c108.png)

15. The `pending` load balancer lb-28 back to `running` and can route correctly
  ![image](https://user-images.githubusercontent.com/29251855/163149098-14e6d182-6cdd-47e5-b161-0b2674117571.png)

**Comment by TachunLin** [comment]
Verified **`fixed`** for KDM release regression test with the following configuration
* Rancher: v2.6.4
* Harvester:  master-03ee6213-head (04/18)
* RKE2 version: v1.23.6-rc3+rke2r1
* Cloud provider version: 0.1.1100 

Create 100 `Pool` load balancers, all of them are running and can be routed correctly
  ![image](https://user-images.githubusercontent.com/29251855/165042864-354b62f7-bbed-44b5-9b07-23bc5e26c580.png)
